File: dp_primer/dp_primer_apples_step2.cs
// 実行時間 20ms
using System;
using System.Linq;

internal class Program
{
    static void Main()
    {
        int[] nxayb = Console.ReadLine().Split().Select( int.Parse ).ToArray();
        int[] p = Enumerable.Repeat( 10000000, nxayb[0] + nxayb[3] ).ToArray();

        p[0] = 0;
        for( int i = 1; i < nxayb[0] + nxayb[3]; i++ ) {
            if( nxayb[1] <= i ) {
                p[i] = Math.Min( p[i], p[i - nxayb[1]] + nxayb[2] );
            }
            if( nxayb[3] <= i ) {
                p[i] = Math.Min( p[i], p[i - nxayb[3]] + nxayb[4] );
            }
        }
        for( int i = nxayb[0] + nxayb[3] - 2; 1 <= i; i-- ) {
            p[i] = Math.Min( p[i], p[i + 1] );
        }
        Console.WriteLine( p[nxayb[0]] );
        return;
    }
}

File: dp_primer/dp_primer_apples_step2.py
# 実行時間 20ms
n, x, a, y, b = list(map(int, input().split()))
p = []

for _ in range(n + y):
    p.append(10000 * 1000)
p[0] = 0
for i in range(1, n + y):
    if x <= i:
        p[i] = min(p[i], p[i - x] + a)
    if y <= i:
        p[i] = min(p[i], p[i - y] + b)
for i in range(n + y - 2, 0, -1):
    p[i] = min(p[i], p[i + 1])
print(p[n])
